File: src/lzo1x_decompress_safe.rs
extern {
    fn memcpy(
        __dest : *mut ::std::os::raw::c_void,
        __src : *const ::std::os::raw::c_void,
        __n : usize
    ) -> *mut ::std::os::raw::c_void;
}

unsafe extern fn get_unaligned_le16(
    mut p : *const ::std::os::raw::c_void
) -> u16 {
    let mut ret : u16 = 0u16;
    memcpy(
        &mut ret as (*mut u16) as (*mut ::std::os::raw::c_void),
        p,
        ::std::mem::size_of::<u16>()
    );
    ret
}

#[no_mangle]
pub unsafe extern fn lzo1x_decompress_safe(
    mut in_ : *const u8,
    mut in_len : usize,
    mut out : *mut u8,
    mut out_len : *mut usize
) -> i32 {
    let mut _currentBlock;
    let mut op : *mut u8;
    let mut ip : *const u8;
    let mut t : usize;
    let mut next : usize;
    let mut state : usize = 0usize;
    let mut m_pos : *const u8;
    let ip_end : *const u8 = in_.offset(in_len as (isize));
    let op_end : *mut u8 = out.offset(*out_len as (isize));
    op = out;
    ip = in_;
    if in_len < 3usize {
        *out_len = ((op as (isize)).wrapping_sub(
                        out as (isize)
                    ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
        -4i32
    } else {
        if *ip as (i32) > 17i32 {
            t = (*{
                      let _old = ip;
                      ip = ip.offset(1isize);
                      _old
                  } as (i32) - 17i32) as (usize);
            if t < 4usize {
                next = t;
                state = next;
                t = next;
                if !(((ip_end as (isize)).wrapping_sub(
                          ip as (isize)
                      ) / ::std::mem::size_of::<u8>(
                          ) as (isize)) as (usize) >= t.wrapping_add(3usize)) {
                    *out_len = ((op as (isize)).wrapping_sub(
                                    out as (isize)
                                ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
                    return -4i32;
                } else if !(((op_end as (isize)).wrapping_sub(
                                 op as (isize)
                             ) / ::std::mem::size_of::<u8>() as (isize)) as (usize) >= t) {
                    *out_len = ((op as (isize)).wrapping_sub(
                                    out as (isize)
                                ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
                    return -6i32;
                } else {
                    'loop12: loop {
                        if !(t > 0usize) {
                            break;
                        }
                        *{
                             let _old = op;
                             op = op.offset(1isize);
                             _old
                         } = *{
                                  let _old = ip;
                                  ip = ip.offset(1isize);
                                  _old
                              };
                        t = t.wrapping_sub(1usize);
                    }
                }
            } else if !(((op_end as (isize)).wrapping_sub(
                             op as (isize)
                         ) / ::std::mem::size_of::<u8>() as (isize)) as (usize) >= t) {
                *out_len = ((op as (isize)).wrapping_sub(
                                out as (isize)
                            ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
                return -6i32;
            } else if !(((ip_end as (isize)).wrapping_sub(
                             ip as (isize)
                         ) / ::std::mem::size_of::<u8>(
                             ) as (isize)) as (usize) >= t.wrapping_add(3usize)) {
                *out_len = ((op as (isize)).wrapping_sub(
                                out as (isize)
                            ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
                return -4i32;
            } else {
                'loop5: loop {
                    *{
                         let _old = op;
                         op = op.offset(1isize);
                         _old
                     } = *{
                              let _old = ip;
                              ip = ip.offset(1isize);
                              _old
                          };
                    if !({
                             t = t.wrapping_sub(1usize);
                             t
                         } > 0usize) {
                        break;
                    }
                }
                state = 4usize;
            }
        }
        'loop13: loop {
            t = *{
                     let _old = ip;
                     ip = ip.offset(1isize);
                     _old
                 } as (usize);
            let mut do_match_next : i32 = 0i32;
            if t < 16usize {
                if state == 0usize {
                    if t == 0usize {
                        let mut offset : usize;
                        let mut ip_last : *const u8 = ip;
                        'loop63: loop {
                            if !(*ip as (i32) == 0i32) {
                                break;
                            }
                            ip = ip.offset(1isize);
                            if !(((ip_end as (isize)).wrapping_sub(
                                      ip as (isize)
                                  ) / ::std::mem::size_of::<u8>(
                                      ) as (isize)) as (usize) >= 1usize) {
                                _currentBlock = 76;
                                break 'loop13;
                            }
                        }
                        offset = ((ip as (isize)).wrapping_sub(
                                      ip_last as (isize)
                                  ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
                        if offset > (!0i32 as (usize)).wrapping_div(255usize).wrapping_sub(
                                        2usize
                                    ) {
                            _currentBlock = 74;
                            break;
                        }
                        offset = (offset << 8i32).wrapping_sub(offset);
                        t = t.wrapping_add(
                                offset.wrapping_add(15usize).wrapping_add(
                                    *{
                                         let _old = ip;
                                         ip = ip.offset(1isize);
                                         _old
                                     } as (usize)
                                )
                            );
                    }
                    t = t.wrapping_add(3usize);
                    if !(((op_end as (isize)).wrapping_sub(
                              op as (isize)
                          ) / ::std::mem::size_of::<u8>() as (isize)) as (usize) >= t) {
                        _currentBlock = 73;
                        break;
                    }
                    if !(((ip_end as (isize)).wrapping_sub(
                              ip as (isize)
                          ) / ::std::mem::size_of::<u8>(
                              ) as (isize)) as (usize) >= t.wrapping_add(3usize)) {
                        _currentBlock = 72;
                        break;
                    }
                    'loop69: loop {
                        *{
                             let _old = op;
                             op = op.offset(1isize);
                             _old
                         } = *{
                                  let _old = ip;
                                  ip = ip.offset(1isize);
                                  _old
                              };
                        if !({
                                 t = t.wrapping_sub(1usize);
                                 t
                             } > 0usize) {
                            break;
                        }
                    }
                    state = 4usize;
                    continue;
                } else if state != 4usize {
                    next = t & 3usize;
                    m_pos = op.offset(-1isize) as (*const u8);
                    m_pos = m_pos.offset(-((t >> 2i32) as (isize)));
                    m_pos = m_pos.offset(
                                -((*{
                                        let _old = ip;
                                        ip = ip.offset(1isize);
                                        _old
                                    } as (i32) << 2i32) as (isize))
                            );
                    if m_pos < out as (*const u8) {
                        _currentBlock = 60;
                        break;
                    }
                    if !(((op_end as (isize)).wrapping_sub(
                              op as (isize)
                          ) / ::std::mem::size_of::<u8>() as (isize)) as (usize) >= 2usize) {
                        _currentBlock = 59;
                        break;
                    }
                    *op.offset(0isize) = *m_pos.offset(0isize);
                    *op.offset(1isize) = *m_pos.offset(1isize);
                    op = op.offset(2isize);
                    do_match_next = 1i32;
                } else {
                    next = t & 3usize;
                    m_pos = op.offset(-((1i32 + 0x800i32) as (isize))) as (*const u8);
                    m_pos = m_pos.offset(-((t >> 2i32) as (isize)));
                    m_pos = m_pos.offset(
                                -((*{
                                        let _old = ip;
                                        ip = ip.offset(1isize);
                                        _old
                                    } as (i32) << 2i32) as (isize))
                            );
                    t = 3usize;
                }
            } else if t >= 64usize {
                next = t & 3usize;
                m_pos = op.offset(-1isize) as (*const u8);
                m_pos = m_pos.offset(-((t >> 2i32 & 7usize) as (isize)));
                m_pos = m_pos.offset(
                            -((*{
                                    let _old = ip;
                                    ip = ip.offset(1isize);
                                    _old
                                } as (i32) << 3i32) as (isize))
                        );
                t = (t >> 5i32).wrapping_sub(1usize).wrapping_add(
                        (3i32 - 1i32) as (usize)
                    );
            } else if t >= 32usize {
                t = (t & 31usize).wrapping_add((3i32 - 1i32) as (usize));
                if t == 2usize {
                    let mut offset : usize;
                    let mut ip_last : *const u8 = ip;
                    'loop30: loop {
                        if !(*ip as (i32) == 0i32) {
                            break;
                        }
                        ip = ip.offset(1isize);
                        if !(((ip_end as (isize)).wrapping_sub(
                                  ip as (isize)
                              ) / ::std::mem::size_of::<u8>() as (isize)) as (usize) >= 1usize) {
                            _currentBlock = 37;
                            break 'loop13;
                        }
                    }
                    offset = ((ip as (isize)).wrapping_sub(
                                  ip_last as (isize)
                              ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
                    if offset > (!0i32 as (usize)).wrapping_div(255usize).wrapping_sub(
                                    2usize
                                ) {
                        _currentBlock = 35;
                        break;
                    }
                    offset = (offset << 8i32).wrapping_sub(offset);
                    t = t.wrapping_add(
                            offset.wrapping_add(31usize).wrapping_add(
                                *{
                                     let _old = ip;
                                     ip = ip.offset(1isize);
                                     _old
                                 } as (usize)
                            )
                        );
                    if !(((ip_end as (isize)).wrapping_sub(
                              ip as (isize)
                          ) / ::std::mem::size_of::<u8>() as (isize)) as (usize) >= 2usize) {
                        _currentBlock = 34;
                        break;
                    }
                }
                m_pos = op.offset(-1isize) as (*const u8);
                next = get_unaligned_le16(
                           ip as (*const ::std::os::raw::c_void)
                       ) as (usize);
                ip = ip.offset(2isize);
                m_pos = m_pos.offset(-((next >> 2i32) as (isize)));
                next = next & 3usize;
            } else {
                m_pos = op as (*const u8);
                m_pos = m_pos.offset(-(((t & 8usize) << 11i32) as (isize)));
                t = (t & 7usize).wrapping_add((3i32 - 1i32) as (usize));
                if t == 2usize {
                    let mut offset : usize;
                    let mut ip_last : *const u8 = ip;
                    'loop18: loop {
                        if !(*ip as (i32) == 0i32) {
                            break;
                        }
                        ip = ip.offset(1isize);
                        if !(((ip_end as (isize)).wrapping_sub(
                                  ip as (isize)
                              ) / ::std::mem::size_of::<u8>() as (isize)) as (usize) >= 1usize) {
                            _currentBlock = 27;
                            break 'loop13;
                        }
                    }
                    offset = ((ip as (isize)).wrapping_sub(
                                  ip_last as (isize)
                              ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
                    if offset > (!0i32 as (usize)).wrapping_div(255usize).wrapping_sub(
                                    2usize
                                ) {
                        _currentBlock = 25;
                        break;
                    }
                    offset = (offset << 8i32).wrapping_sub(offset);
                    t = t.wrapping_add(
                            offset.wrapping_add(7usize).wrapping_add(
                                *{
                                     let _old = ip;
                                     ip = ip.offset(1isize);
                                     _old
                                 } as (usize)
                            )
                        );
                    if !(((ip_end as (isize)).wrapping_sub(
                              ip as (isize)
                          ) / ::std::mem::size_of::<u8>() as (isize)) as (usize) >= 2usize) {
                        _currentBlock = 24;
                        break;
                    }
                }
                next = get_unaligned_le16(
                           ip as (*const ::std::os::raw::c_void)
                       ) as (usize);
                ip = ip.offset(2isize);
                m_pos = m_pos.offset(-((next >> 2i32) as (isize)));
                next = next & 3usize;
                if m_pos == op as (*const u8) {
                    _currentBlock = 23;
                    break;
                }
                m_pos = m_pos.offset(-0x4000isize);
            }
            if do_match_next == 0 {
                if m_pos < out as (*const u8) {
                    _currentBlock = 58;
                    break;
                }
                let mut oe : *mut u8 = op.offset(t as (isize));
                if !(((op_end as (isize)).wrapping_sub(
                          op as (isize)
                      ) / ::std::mem::size_of::<u8>() as (isize)) as (usize) >= t) {
                    _currentBlock = 57;
                    break;
                }
                *op.offset(0isize) = *m_pos.offset(0isize);
                *op.offset(1isize) = *m_pos.offset(1isize);
                op = op.offset(2isize);
                m_pos = m_pos.offset(2isize);
                'loop49: loop {
                    *{
                         let _old = op;
                         op = op.offset(1isize);
                         _old
                     } = *{
                              let _old = m_pos;
                              m_pos = m_pos.offset(1isize);
                              _old
                          };
                    if !(op < oe) {
                        break;
                    }
                }
            }
            state = next;
            t = next;
            if !(((ip_end as (isize)).wrapping_sub(
                      ip as (isize)
                  ) / ::std::mem::size_of::<u8>(
                      ) as (isize)) as (usize) >= t.wrapping_add(3usize)) {
                _currentBlock = 56;
                break;
            }
            if !(((op_end as (isize)).wrapping_sub(
                      op as (isize)
                  ) / ::std::mem::size_of::<u8>() as (isize)) as (usize) >= t) {
                _currentBlock = 55;
                break;
            }
            'loop53: loop {
                if !(t > 0usize) {
                    break;
                }
                *{
                     let _old = op;
                     op = op.offset(1isize);
                     _old
                 } = *{
                          let _old = ip;
                          ip = ip.offset(1isize);
                          _old
                      };
                t = t.wrapping_sub(1usize);
            }
        }
        if _currentBlock == 23 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            if t != 3usize {
                -1i32
            } else if ip == ip_end {
                0i32
            } else if ip < ip_end {
                -8i32
            } else {
                -4i32
            }
        } else if _currentBlock == 24 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -4i32
        } else if _currentBlock == 25 {
            -1i32
        } else if _currentBlock == 27 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -4i32
        } else if _currentBlock == 34 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -4i32
        } else if _currentBlock == 35 {
            -1i32
        } else if _currentBlock == 37 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -4i32
        } else if _currentBlock == 55 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -6i32
        } else if _currentBlock == 56 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -4i32
        } else if _currentBlock == 57 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -6i32
        } else if _currentBlock == 58 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -6i32
        } else if _currentBlock == 59 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -6i32
        } else if _currentBlock == 60 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -6i32
        } else if _currentBlock == 72 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -4i32
        } else if _currentBlock == 73 {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -6i32
        } else if _currentBlock == 74 {
            -1i32
        } else {
            *out_len = ((op as (isize)).wrapping_sub(
                            out as (isize)
                        ) / ::std::mem::size_of::<u8>() as (isize)) as (usize);
            -4i32
        }
    }
}

File: src/lzo1x_decompress_safe.c
/*
 *  LZO1X Decompressor from LZO
 *
 *  Copyright (C) 1996-2012 Markus F.X.J. Oberhumer <markus@oberhumer.com>
 *
 *  The full LZO package can be found at:
 *  http://www.oberhumer.com/opensource/lzo/
 *
 *  Changed for Linux kernel use by:
 *  Nitin Gupta <nitingupta910@gmail.com>
 *  Richard Purdie <rpurdie@openedhand.com>
 */

#include <string.h>
#include <stdlib.h>
#include <stdint.h>
typedef uint32_t u32;
typedef uint16_t u16;
typedef uint64_t u64;
#include "lzodefs.h"
#include "lzo.h"

#define unlikely(x) (x)
#define likely(x) (x)
#define noinline

static inline u32 get_unaligned_le32(const void *p)
{
        u32 ret;
        memcpy(&ret, p, sizeof(ret));
        return ret;
}

static inline u16 get_unaligned_le16(const void *p)
{
        u16 ret = 0;
        memcpy(&ret, p, sizeof(ret));
        return ret;
}

static inline u32 get_unaligned(const void *p)
{
        u32 ret;
        memcpy(&ret, p, sizeof(ret));
        return ret;
}


#define HAVE_IP(x)      ((size_t)(ip_end - ip) >= (size_t)(x))
#define HAVE_OP(x)      ((size_t)(op_end - op) >= (size_t)(x))
#define NEED_IP(x)      if (!HAVE_IP(x)) goto input_overrun
#define NEED_OP(x)      if (!HAVE_OP(x)) goto output_overrun
#define TEST_LB(m_pos)  if ((m_pos) < out) goto lookbehind_overrun

/* This MAX_255_COUNT is the maximum number of times we can add 255 to a base
 * count without overflowing an integer. The multiply will overflow when
 * multiplying 255 by more than MAXINT/255. The sum will overflow earlier
 * depending on the base count. Since the base count is taken from a u8
 * and a few bits, it is safe to assume that it will always be lower than
 * or equal to 2*255, thus we can always prevent any overflow by accepting
 * two less 255 steps. See Documentation/lzo.txt for more information.
 */
#define MAX_255_COUNT      ((((size_t)~0) / 255) - 2)

int lzo1x_decompress_safe(const unsigned char *in, size_t in_len,
			  unsigned char *out, size_t *out_len)
{
	unsigned char *op;
	const unsigned char *ip;
	size_t t, next;
	size_t state = 0;
	const unsigned char *m_pos;
	const unsigned char * const ip_end = in + in_len;
	unsigned char * const op_end = out + *out_len;

	op = out;
	ip = in;

	if (unlikely(in_len < 3)) {
		*out_len = op - out;
		return LZO_E_INPUT_OVERRUN;
	}

	if (*ip > 17) {
		t = *ip++ - 17;
		if (t < 4) {
			next = t;
			state = next;
			t = next;
			{
				NEED_IP(t + 3);
				NEED_OP(t);
				while (t > 0) {
					*op++ = *ip++;
					t--;
				}
			}
		} else {
			{
				NEED_OP(t);
				NEED_IP(t + 3);
				do {
					*op++ = *ip++;
				} while (--t > 0);
			}
			state = 4;
		}
	}

	for (;;) {
		t = *ip++;
		int do_match_next = 0;
		if (t < 16) {
			if (likely(state == 0)) {
				if (unlikely(t == 0)) {
					size_t offset;
					const unsigned char *ip_last = ip;

					while (unlikely(*ip == 0)) {
						ip++;
						NEED_IP(1);
					}
					offset = ip - ip_last;
					if (unlikely(offset > MAX_255_COUNT))
						return LZO_E_ERROR;

					offset = (offset << 8) - offset;
					t += offset + 15 + *ip++;
				}
				t += 3;
copy_literal_run:
				{
					NEED_OP(t);
					NEED_IP(t + 3);
					do {
						*op++ = *ip++;
					} while (--t > 0);
				}
				state = 4;
				continue;
			} else if (state != 4) {
				next = t & 3;
				m_pos = op - 1;
				m_pos -= t >> 2;
				m_pos -= *ip++ << 2;
				TEST_LB(m_pos);
				NEED_OP(2);
				op[0] = m_pos[0];
				op[1] = m_pos[1];
				op += 2;
				do_match_next = 1;
			} else {
				next = t & 3;
				m_pos = op - (1 + M2_MAX_OFFSET);
				m_pos -= t >> 2;
				m_pos -= *ip++ << 2;
				t = 3;
			}
		} else if (t >= 64) {
			next = t & 3;
			m_pos = op - 1;
			m_pos -= (t >> 2) & 7;
			m_pos -= *ip++ << 3;
			t = (t >> 5) - 1 + (3 - 1);
		} else if (t >= 32) {
			t = (t & 31) + (3 - 1);
			if (unlikely(t == 2)) {
				size_t offset;
				const unsigned char *ip_last = ip;

				while (unlikely(*ip == 0)) {
					ip++;
					NEED_IP(1);
				}
				offset = ip - ip_last;
				if (unlikely(offset > MAX_255_COUNT))
					return LZO_E_ERROR;

				offset = (offset << 8) - offset;
				t += offset + 31 + *ip++;
				NEED_IP(2);
			}
			m_pos = op - 1;
			next = get_unaligned_le16(ip);
			ip += 2;
			m_pos -= next >> 2;
			next &= 3;
		} else {
			m_pos = op;
			m_pos -= (t & 8) << 11;
			t = (t & 7) + (3 - 1);
			if (unlikely(t == 2)) {
				size_t offset;
				const unsigned char *ip_last = ip;

				while (unlikely(*ip == 0)) {
					ip++;
					NEED_IP(1);
				}
				offset = ip - ip_last;
				if (unlikely(offset > MAX_255_COUNT))
					return LZO_E_ERROR;

				offset = (offset << 8) - offset;
				t += offset + 7 + *ip++;
				NEED_IP(2);
			}
			next = get_unaligned_le16(ip);
			ip += 2;
			m_pos -= next >> 2;
			next &= 3;
			if (m_pos == op) {
				*out_len = op - out;
				return (t != 3       ? LZO_E_ERROR :
					ip == ip_end ? LZO_E_OK :
					ip <  ip_end ? LZO_E_INPUT_NOT_CONSUMED : LZO_E_INPUT_OVERRUN);

			}
			m_pos -= 0x4000;
		}
		if (!do_match_next) {
		TEST_LB(m_pos);
		{
			unsigned char *oe = op + t;
			NEED_OP(t);
			op[0] = m_pos[0];
			op[1] = m_pos[1];
			op += 2;
			m_pos += 2;
			do {
				*op++ = *m_pos++;
			} while (op < oe);
		}
		}
match_next:
		state = next;
		t = next;
		{
			NEED_IP(t + 3);
			NEED_OP(t);
			while (t > 0) {
				*op++ = *ip++;
				t--;
			}
		}
	}

	*out_len = op - out;
	return (t != 3       ? LZO_E_ERROR :
		ip == ip_end ? LZO_E_OK :
		ip <  ip_end ? LZO_E_INPUT_NOT_CONSUMED : LZO_E_INPUT_OVERRUN);

input_overrun:
	*out_len = op - out;
	return LZO_E_INPUT_OVERRUN;

output_overrun:
	*out_len = op - out;
	return LZO_E_OUTPUT_OVERRUN;

lookbehind_overrun:
	*out_len = op - out;
	return LZO_E_LOOKBEHIND_OVERRUN;
}
